packages/react-reconciler/src/ReactFiberWorkLoop.new.js

```diff
@@ -2898,64 +2905,72 @@ function flushRenderPhaseStrictModeWarningsInDEV() {
   }
 }
 
-function commitDoubleInvokeEffectsInDEV(
+function recursivelyTraverseAndDoubleInvokeEffectsInDEV(
+  root: FiberRoot,
   fiber: Fiber,
   hasPassiveEffects: boolean,
+  doubleInvokeEffects: boolean,
 ) {
-  if (__DEV__ && enableStrictEffects) {
-    // TODO (StrictEffects) Should we set a marker on the root if it contains strict effects
-    // so we don't traverse unnecessarily? similar to subtreeFlags but just at the root level.
-    // Maybe not a big deal since this is DEV only behavior.
-
-    setCurrentDebugFiberInDEV(fiber);
-    invokeEffectsInDev(fiber, MountLayoutDev, invokeLayoutEffectUnmountInDEV);
-    if (hasPassiveEffects) {
-      invokeEffectsInDev(
-        fiber,
-        MountPassiveDev,
-        invokePassiveEffectUnmountInDEV,
-      );
-    }
+  if (fiber.type === REACT_STRICT_MODE_TYPE) {
+    doubleInvokeEffects = true;
```

Comment:
Because of the way you've structured this loop, this will be set to `true` when you traverse the siblings, too, which is not correct. It should only be true when traversing the children.

I find it's easier to track this kind of recursive logic if you split the traversal into two separate functions, something like this:

```js
function commitDoubleInvokeEffectsInDEV(root, fiber, parentIsInStrictMode) {
  if (fiber.flags & PlacementDEV) {
    // Call disappear/reappear, disconnect/reconnect, etc
    // Don't need to keep traversing
  } else {
    // Keep searching for placements
    const isInStrictMode = parentIsInStrictMode || isStrictModeFiber;
    recursivelyTraverseAndDoubleInvokeEffectsInDEV(root, fiber, isInStrictMode);
  }
}

function recursivelyTraverseAndDoubleInvokeEffectsInDEV(
  root,
  parentFiber,
  isInStrictMode,
) {
  if (parentFiber.subtreeFlags & PlacementDEV) {
    let child = parentFiber.child;
    while (child !== null) {
      commitDoubleInvokeEffectsInDEV(root, child, isInStrictMode);
      child = child.sibling;
    }
  }
}
```

Reply:
Thanks. This is much better.